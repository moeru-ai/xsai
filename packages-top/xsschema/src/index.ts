export { toJsonSchema, toJsonSchema as toJSONSchema, toJsonSchemaSync, toJsonSchemaSync as toJSONSchemaSync } from './to-json-schema'
export type { Infer, InferIn, JSONSchema as JsonSchema, JSONSchema, Schema, StandardSchemaVendor } from './types'
export { validate } from './validate'
export { registerStandardSchemaVendor, registerStandardSchemaVendors } from './vendors'
