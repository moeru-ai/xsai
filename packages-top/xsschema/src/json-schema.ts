import type { JsonSchema } from './types'

export const jsonSchema = (schema: JsonSchema) => schema

export const strictJsonSchema = (schema: JsonSchema): JsonSchema => ({
  ...schema,
  additionalProperties: false,
  properties: schema.properties != null
    ? Object.fromEntries(
        Object.entries(schema.properties)
          .map(([k, v]) => [
            k,
            v != null && typeof v === 'object' && 'type' in v && v.type === 'object'
              ? strictJsonSchema(v)
              : v,
          ]),
      )
    : schema.properties,
})
