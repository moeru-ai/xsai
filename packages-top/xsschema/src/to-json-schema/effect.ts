import type { JSONSchema7 } from 'json-schema'

export const toJsonSchema = async () => {
  try {
    const { JSONSchema } = await import('effect')
    // eslint-disable-next-line @masknet/type-prefer-return-type-annotation
    return JSONSchema.make as (schema: unknown) => JSONSchema7
  }
  catch {
    throw new Error('xsschema: Missing dependencies "effect"')
  }
}
