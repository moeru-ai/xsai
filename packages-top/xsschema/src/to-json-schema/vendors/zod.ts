import type { ZodSchema } from 'zod'
import type { ZodType } from 'zod4'

import type { ToJsonSchemaFn } from '.'

export const getToJsonSchemaFn = async (): Promise<ToJsonSchemaFn> => {
  let toJSONSchema: ToJsonSchemaFn = (_schema: unknown) => {
    throw new Error('xsschema: Missing dependencies "zod" or "@zod/mini".')
  }
  let zodToJSONSchema: ToJsonSchemaFn = (_schema: unknown) => {
    throw new Error('xsschema: Missing dependencies "zod-to-json-schema".')
  }

  try {
    const { z } = await import('@zod/mini')
    toJSONSchema = z.toJSONSchema as ToJsonSchemaFn
  }
  catch {}

  try {
    const { z } = await import('zod')
    if ('toJSONSchema' in z)
      toJSONSchema = z.toJSONSchema as ToJsonSchemaFn
  }
  catch (err) {
    if (err instanceof Error)
      console.error(err.message)
  }

  try {
    const { zodToJsonSchema } = await import('zod-to-json-schema')
    zodToJSONSchema = zodToJsonSchema as ToJsonSchemaFn
  }
  catch (err) {
    if (err instanceof Error)
      console.error(err.message)
  }

  return async (schema: unknown) => {
    if ('_zod' in (schema as ZodSchema | ZodType))
      return toJSONSchema(schema)
    else
      return zodToJSONSchema(schema)
  }
}
