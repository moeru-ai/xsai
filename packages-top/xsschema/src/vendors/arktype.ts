import type { StandardSchemaV1 } from '@standard-schema/spec'
import type { Type } from 'arktype'
import type { JSONSchema7 } from 'json-schema'

export const toJsonSchema = (schema: StandardSchemaV1): JSONSchema7 =>
  (schema as Type).toJsonSchema() as JSONSchema7
