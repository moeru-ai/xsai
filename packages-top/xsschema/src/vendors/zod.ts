import type { StandardSchemaV1 } from '@standard-schema/spec'
import type { JSONSchema7 } from 'json-schema'
import type { ZodSchema } from 'zod'

import zodToJsonSchemaImpl from 'zod-to-json-schema'

export const toJsonSchema = (schema: StandardSchemaV1): JSONSchema7 => {
  // eslint-disable-next-line @masknet/type-prefer-return-type-annotation
  return zodToJsonSchemaImpl(schema as ZodSchema<any>) as JSONSchema7
}
