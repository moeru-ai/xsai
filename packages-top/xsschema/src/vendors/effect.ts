import type { StandardSchemaV1 } from '@standard-schema/spec'
import type { Schema } from 'effect'
import type { JSONSchema7 } from 'json-schema'

import { JSONSchema } from 'effect'

export const toJsonSchema = (schema: StandardSchemaV1): JSONSchema7 =>
  // eslint-disable-next-line @masknet/type-no-force-cast-via-top-type
  JSONSchema.make(schema as unknown as Schema.Schema<any, any, any>) as JSONSchema7
