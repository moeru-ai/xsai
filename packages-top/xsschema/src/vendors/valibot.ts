import type { StandardSchemaV1 } from '@standard-schema/spec'
import type { JSONSchema7 } from 'json-schema'
import type { BaseIssue, BaseSchema } from 'valibot'

import { toJsonSchema as toJsonSchemaImpl } from '@valibot/to-json-schema'

export const toJsonSchema = (schema: StandardSchemaV1): JSONSchema7 => {
  return toJsonSchemaImpl(schema as BaseSchema<unknown, unknown, BaseIssue<unknown>>)
}
