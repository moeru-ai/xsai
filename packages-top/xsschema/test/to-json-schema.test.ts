import { type } from 'arktype'
import { Schema } from 'effect'
import * as v from 'valibot'
import { describe, expect, it } from 'vitest'
import * as z from 'zod'

import { toJsonSchema } from '../src'

describe('toJsonSchema', () => {
  it('arktype', async () => {
    const schema = type({
      myString: 'string',
      myUnion: 'number | boolean',
    }).describe('My neat object schema')

    const jsonSchema = await toJsonSchema(schema)
    expect(jsonSchema).toMatchSnapshot()
  })

  it('effect', async () => {
    const schema = Schema.Struct({
      myString: Schema.String,
      myUnion: Schema.Union(Schema.Number, Schema.Boolean),
    }).annotations({ description: 'My neat object schema' })

    // https://github.com/Effect-TS/effect/issues/4494
    const jsonSchema = await toJsonSchema(Object.assign(schema, Schema.standardSchemaV1(schema)))
    expect(jsonSchema).toMatchSnapshot()
  })

  it('valibot', async () => {
    const schema = v.pipe(
      v.object({
        myString: v.string(),
        myUnion: v.union([v.number(), v.boolean()]),
      }),
      v.description('My neat object schema'),
    )

    const jsonSchema = await toJsonSchema(schema)
    expect(jsonSchema).toMatchSnapshot()
  })

  it('zod', async () => {
    const schema = z.object({
      myString: z.string(),
      myUnion: z.union([z.number(), z.boolean()]),
    }).describe('My neat object schema')

    const jsonSchema = await toJsonSchema(schema)
    expect(jsonSchema).toMatchSnapshot()
  })
})
