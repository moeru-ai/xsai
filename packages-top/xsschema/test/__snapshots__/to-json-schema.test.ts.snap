// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toJsonSchema > arktype 1`] = `
{
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;

exports[`toJsonSchema > effect 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;

exports[`toJsonSchema > valibot 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "anyOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;

exports[`toJsonSchema > zod 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "type": [
        "number",
        "boolean",
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;
